<a href="https://colab.research.google.com/github/suryakarasala/Build-a-Regression-Model-in-Keras/blob/main/regression_model(D).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **WEEK 5 Assignment – Introduction to Deep Learning & Neural Networks with Keras**
Surya Prakash Karasala
 
Date: 03/22/2021

## **Importing Required Packages**

In [1]:
import pandas as pd
import numpy as np

We will be using the dataset provided in the assignment

The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:

1. Cement

2. Blast Furnace Slag

3. Fly Ash

4. Water

5. Superplasticizer

6. Coarse Aggregate

7. Fine Aggregate

Let's read the dataset into a pandas dataframe.

## Loading Data

In [2]:
concrete_data = pd.read_csv('https://cocl.us/concrete_data')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Fine Aggregate  Age  Strength
0   540.0                 0.0      0.0  ...           676.0   28     79.99
1   540.0                 0.0      0.0  ...           676.0   28     61.89
2   332.5               142.5      0.0  ...           594.0  270     40.27
3   332.5               142.5      0.0  ...           594.0  365     41.05
4   198.6               132.4      0.0  ...           825.5  360     44.30

[5 rows x 9 columns]

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa.

###**Let's check how many data points we have**.

In [3]:
concrete_data.shape

(1030, 9)


So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.

Let's check the dataset for any missing values.

In [4]:
concrete_data.describe()

Cement  Blast Furnace Slag  ...          Age     Strength
count  1030.000000         1030.000000  ...  1030.000000  1030.000000
mean    281.167864           73.895825  ...    45.662136    35.817961
std     104.506364           86.279342  ...    63.169912    16.705742
min     102.000000            0.000000  ...     1.000000     2.330000
25%     192.375000            0.000000  ...     7.000000    23.710000
50%     272.900000           22.000000  ...    28.000000    34.445000
75%     350.000000          142.950000  ...    56.000000    46.135000
max     540.000000          359.400000  ...   365.000000    82.600000

[8 rows x 9 columns]

In [5]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.

### **Cleaning and Normalizing Data**
The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [6]:
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

Let's do a quick sanity check of the predictors and the target dataframes.

In [7]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Coarse Aggregate  Fine Aggregate  Age
0   540.0                 0.0      0.0  ...            1040.0           676.0   28
1   540.0                 0.0      0.0  ...            1055.0           676.0   28
2   332.5               142.5      0.0  ...             932.0           594.0  270
3   332.5               142.5      0.0  ...             932.0           594.0  365
4   198.6               132.4      0.0  ...             978.4           825.5  360

[5 rows x 8 columns]

In [8]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.

In [9]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag  ...  Fine Aggregate       Age
0  2.476712           -0.856472  ...       -1.217079 -0.279597
1  2.476712           -0.856472  ...       -1.217079 -0.279597
2  0.491187            0.795140  ...       -2.239829  3.551340
3  0.491187            0.795140  ...       -2.239829  5.055221
4 -0.790075            0.678079  ...        0.647569  4.976069

[5 rows x 8 columns]

In [10]:
n_cols = predictors_norm.shape[1] # number of predictors
n_cols

8

# **Import Keras**
Let's go ahead and import the Keras library

In [11]:
import keras

As you can see, the TensorFlow backend was used to install the Keras library.

Let's import the rest of the packages from the Keras library that we will need to build our regressoin model

In [12]:
from keras.models import Sequential
from keras.layers import Dense

In [13]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function creates a model that has three hidden layers each with 10 neurons and ReLU activation function. It uses the adam optimizer and the mean squared error as the loss function.

Let's import scikit-learn in order to randomly split the data into a training and test sets

In [15]:
from sklearn.model_selection import train_test_split

Splitting the data into a training and test sets by holding 30% of the data for testing

In [16]:
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=42)

## **Train and Test the Network**
Let's call the function now to create our model.

In [17]:
# build the model
model = regression_model()

Next, we will train the model for 50 epochs.

In [18]:
# fit the model
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=2)

Epoch 1/50
23/23 - 1s - loss: 1571.1720
Epoch 2/50
23/23 - 0s - loss: 1543.7062
Epoch 3/50
23/23 - 0s - loss: 1505.5045
Epoch 4/50
23/23 - 0s - loss: 1445.9047
Epoch 5/50
23/23 - 0s - loss: 1355.7817
Epoch 6/50
23/23 - 0s - loss: 1225.1033
Epoch 7/50
23/23 - 0s - loss: 1045.6276
Epoch 8/50
23/23 - 0s - loss: 827.0839
Epoch 9/50
23/23 - 0s - loss: 610.2446
Epoch 10/50
23/23 - 0s - loss: 433.6354
Epoch 11/50
23/23 - 0s - loss: 320.0462
Epoch 12/50
23/23 - 0s - loss: 259.6555
Epoch 13/50
23/23 - 0s - loss: 228.9420
Epoch 14/50
23/23 - 0s - loss: 212.0455
Epoch 15/50
23/23 - 0s - loss: 201.4913
Epoch 16/50
23/23 - 0s - loss: 193.8458
Epoch 17/50
23/23 - 0s - loss: 187.7815
Epoch 18/50
23/23 - 0s - loss: 182.1181
Epoch 19/50
23/23 - 0s - loss: 177.8428
Epoch 20/50
23/23 - 0s - loss: 173.6141
Epoch 21/50
23/23 - 0s - loss: 169.8841
Epoch 22/50
23/23 - 0s - loss: 166.6092
Epoch 23/50
23/23 - 0s - loss: 163.6473
Epoch 24/50
23/23 - 0s - loss: 160.9434
Epoch 25/50
23/23 - 0s - loss: 158.7883
Ep

Next we need to evaluate the model on the test data.

In [19]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

10/10 [==============================] - 0s 1ms/step - loss: 126.8334


126.83341217041016


Now we need to compute the mean squared error between the predicted concrete strength and the actual concrete strength.

Let's import the mean_squared_error function from Scikit-learn.

In [20]:
from sklearn.metrics import mean_squared_error

In [21]:
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

126.83341566858341 0.0


Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors

In [23]:
total_mean_squared_errors = 50
epochs = 50
mean_squared_errors = []
for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Below is the mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors with normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 23.716758728027344
MSE 2: 25.05706024169922
MSE 3: 21.91568374633789
MSE 4: 23.830663681030273
MSE 5: 25.21424674987793
MSE 6: 25.44489288330078
MSE 7: 27.623306274414062
MSE 8: 20.20354652404785
MSE 9: 23.85636329650879
MSE 10: 24.3265323638916
MSE 11: 21.43800163269043
MSE 12: 21.13308334350586
MSE 13: 24.575210571289062
MSE 14: 24.635398864746094
MSE 15: 25.347064971923828
MSE 16: 15.863798141479492
MSE 17: 20.599393844604492
MSE 18: 23.965030670166016
MSE 19: 23.15103530883789
MSE 20: 24.019250869750977
MSE 21: 22.174333572387695
MSE 22: 23.13707160949707
MSE 23: 19.344417572021484
MSE 24: 21.400218963623047
MSE 25: 25.129375457763672
MSE 26: 25.1956844329834
MSE 27: 19.707443237304688
MSE 28: 22.92325210571289
MSE 29: 23.305309295654297
MSE 30: 23.22614860534668
MSE 31: 18.59229850769043
MSE 32: 21.481067657470703
MSE 33: 20.568294525146484
MSE 34: 21.099760055541992
MSE 35: 23.924991607666016
MSE 36: 26.150367736816406
MSE 37: 18.108060836791992
MSE 38: 24.874481201171875
